### ISSUE TITLE
Missile Rebalancing

### ISSUE THREAD

**Issue by onpon4** [naev/naev#859]
I'm thinking a bit about the need in Naev to rebalance the way missiles work, especially as it pertains to medium ships. I honestly have no idea where to start with this, so I'd like to start a discussion and see if anyone has any thoughts.

Particular issues I have in mind:

1. Missiles are far more useful for A.I. players than for the human player. This was briefly addressed in #832; it's because A.I. players always warp in with maximum ammo capacity whereas players don't necessarily have full ammunition when they warp in. This is a pretty minor issue that could be solved by missile auto-reloading and a mechanism for randomizing how many missiles any given A.I. ship has when it jumps in from an outer system. I think everyone would be in favor of such a change.

2. Missiles are most useful for small, fast craft which have a sufficient turn rate to get the lock-ons missile weapons need and have a particular need, and capability, to stay out of the range of larger ships' firepower.

3. Medium and large ships really don't have any way to counter missiles. Missiles can't be shot down, only jammed, and jamming is really only effective with a really fast ship (or a reasonably fast ship with slow missiles). So in practice, when piloting a medium or large ship, your only possible recourse if pelted with missiles is to hope that the ships attacking you run out of ammo before you get destroyed. This is especially a problem for medium ships.

4. Because of that dynamic, missiles create a strange situation where in practice, medium ships have no credible advantage over light ships. They are too easily pelted into smithereens by pretty much any kind of lock-on missile.

5. At the same time, even given a solution to #1, countering with your own missiles is not particularly effective because small ships have an easier time dodging them and because you're usually outnumbered by a coordinated group attack (#832 addresses the second half of this, but the first half is a consequence of the fundamental way missiles are designed).

Thinking about it, the whole system is so broken that I wouldn't be too upset if we got rid of the missile system entirely. I don't think that would be optimal, however. Missiles (when done right) can be a major enhancement to the experience. I know I've had a lot of fun with missiles in 3-D space shooters for instance.

I'll post my own thoughts in a separate comment since they're disorganized and I don't feel at all sure about them. Anyone else who has any thoughts or ideas, please let them be known! :)

**Comment by onpon4** [comment]
My thoughts:

* Maybe missiles could be reconfigured to only be usable by medium and large ships? But that could create a bad power balance in the opposite direction.
* Maybe missiles could have their range shortened? But that would make them a lot less useful.
* Maybe remove the lock-on requirement?
* Maybe we need some sort of "missile countermeasures" feature that costs some battery or something but destroys incoming missiles? But it's hard to say how to make a feature like that integrate into Naev well.
** Maybe automatically drain the battery if available rather than doing the missile's normal effect?
* Maybe make missiles so slow that they're only effective on large ships? But then that would just shift the main harm missiles cause to large ships.

**Comment by ids1024** [comment]
> Thinking about it, the whole system is so broken that I wouldn't be too upset if we got rid of the missile system entirely. I don't think that would be optimal, however. Missiles (when done right) can be a major enhancement to the experience. 

Yep. It's probably best to keep missiles, even if we change how they work. It's good to have different kinds of weapons with different advantages, and "missiles" of some form are a logical category to include, however we want them to work.

> Because of that dynamic, missiles create a strange situation where in practice, medium ships have no credible advantage over light ships.

The overall balancing between ship classes seems sub-optimal overall, I think. Perhaps I'm not very good at the game, but small ships seem worse than they should be as well. A good fighter should be a very capable ship for many purposes. But if I pilot one though space with high piracy, it seems to require good luck as much as anything not to be blown up, since it doesn't take much to destroy the ship and a swarm of pirates with similarly equipped ships soon shows up.

A small ship can do better at this by equipping good engines and afterburners so it's faster than any other ship, and can at least run away (though it's still easily destroyed). A large ship can just have excessive shielding and armor, plus turrets, to handle a significant attack. I'm not sure how a medium ship fits in here.

I think part of the solution here is probably for pirate AIs to not be particularly interested in attacking a small non-hostile ship without cargo. Which is off topic for this specific issue, but perhaps it's worth considering the broader system.

**TLDR**: Maybe the key here is to figure out exactly what the purpose of each ship class should be, and how to balance it so that each is useful for a different role. Changes to the missile system will probably follow from deciding such questions (like how suitable one class of ship should be for attacking another class).

**Comment by onpon4** [comment]
Indeed, right now there's not that much thought given as to how each class should function, what the advantages and disadvantages should be. Right now it's just a "speed vs power" trade-off which lends play styles that work toward either absolute speed with minimal power, or absolute power with minimal speed, and everything else is just kind of stuck in the middle.

I did start to try to address that back when I implemented the whole time variation thing, which was actually designed to make it less painful to wait for a larger ship to make it from point A to point B, but it occurred to me that it could also serve as an advantage for smaller ships (hence the Hyena and Collective Drone experiencing time slowdowns).

This brings up an interesting point, actually: missiles can serve very well as a way of power balancing smaller ships against larger ships if done right, which is just another good reason to keep them. It seems to me that this would entail making it so that medium ships are immune, or effectively immune, to the destruction that missiles cause, perhaps by making all missiles either very slow (e.g. Caesar missiles) or very short-range (e.g. Mace rockets; some others could be modified to be similarly short-range).

Then there's the question of preventing large ships from being left out of the loop. This wouldn't work right away, but one idea that comes to mind is to strongly encourage the player to equip large ships with fighter bays. This could tie into the idea of making launchers require a specialized slot that can't be used by other weapons. This in combination with fixing up the launched fighter AI and how the use of fighters works could be very effective if it allows capable large ships to defend themselves with long-range attacks by use of fast-moving fighters, while at the same time leaving the smaller ships a fighting chance (even if it's difficult to fight swarms of small fighters all at once).

Hrm, that still doesn't feel satisfactory to me.

**Comment by onpon4** [comment]
I think I can see missiles in two potential places in all of this:

1. As equalizers for smaller ships against larger ships; or
2. As shorter-range missiles used for tactical reasons.

I tend to favor #1 but I'm having a hard time thinking of a way to do that (especially given the hierarchal nature of what ship is allowed to equip what outfits) without creating a power dichotomy between the smallest and the largest anyway.

**Comment by PhoenixRiver** [comment]
> 1. Missiles are far more useful for A.I. players than for the human player.
> This was briefly addressed in #832; it's because A.I. players always warp in
> with maximum ammo capacity whereas players don't necessarily have full
> ammunition when they warp in. This is a pretty minor issue that could be
> solved by missile auto-reloading and a mechanism for randomizing how many
> missiles any given A.I. ship has when it jumps in from an outer system. I
> think everyone would be in favor of such a change.

Bearing in mind this should not be done for ships that are or might be
chasing the player through a system warp... their number of missiles
should remain consistant through the player experience.

> 3. Medium and large ships really don't have any way to counter missiles.
> Missiles can't be shot down, only jammed, and jamming is really only
> effective with a really fast ship (or a reasonably fast ship with slow
> missiles). So in practice, when piloting a medium or large ship, your only
> possible recourse if pelted with missiles is to hope that the ships
> attacking you run out of ammo before you get destroyed. This is especially a
> problem for medium ships.

There should be point defences; these are weapons that are ammo-less
(maybe they are mini-railguns that collect ammo from space dust and
fire it at a significant portion f the speed of light or lasers or
whatever) and with terrible range, they are highly effective against
missiles but almost ineffetive against ships (even small ones, but
maybe semi-effective against drones?) they are only effective when you
have a few of them but that few weighs more than a jammer but less
than serious weapons (ditto with spae requirements); each extra point
defence adds less defence so there is always a chance a missile can
get through (e.g. 1 point defence reduces the number of missiels that
can hit by half, the second reduces the number of missiels to a third,
the next to a quarter, the next to a fifth etc.).

In addition to this the jammer should be scaled to the size of the
ship, the smaller the ship the more effective the jammer is, range and
ability; small ships will likely never both with them since they are
fast enough already, large ships can soak the damage, medium ships
have the advantage in using them.

Combining these two things makes medium ships much harder to kill with
missiles while big ships won't bother with jammers or much point
defences since they can soak the damage and want the space and weight
for other heavier weaponry. Balancing these ight take a little time
but there is a reason that point defences are used in the real world,
missiles are indeed OP; just look at ICBMs, Cruise missiles and the
profusion of Surface to Air, Air to Surface and Air to Air missile
systems that exist; only ships really have point defences except for
the experiemental air lasers for ICBMs and chaff for planes (I guess
that would also be something that could be added, but, again, more
effective for smaller ships and less effective for larger ships - but
would definitely require ammo and not be 100% effective).

> Thinking about it, the whole system is so broken that I wouldn't be too
> upset if we got rid of the missile system entirely. I don't think that would
> be optimal, however. Missiles (when done right) can be a major enhancement
> to the experience. I know I've had a lot of fun with missiles in 3-D space
> shooters for instance.

I also wouldn't be sad to see missile systems removed entirely since a
civilisation with the technology to have effective ship to ship laser
combat means that targeting of missiles by those lasers would make
missiles pointless except in insanely dense swarms that would either
be ineffective of doing serious damage, resisting jamming or avoiding
point defences or would take up so much space as to only allow one or
two volleys by small ships making them incapable of taking down medium
or large ships and still being ineffective at catching small ones...
now, if you switched them to being mines, that'd be a different story,
they would be much more effective being almost entirely invisible,
even if they are unguided (except at the last second maybe) and would
gain the benefit being scaled by ship size, large ships will hit more
due to lack of manouveurability and medium ships less with small ships
hitting almost none...

Just my two cents! :)

Phoenix

**Comment by onpon4** [comment]
Point defense would certainly be really nice. I can't think of a way to implement that given how the game works though (I don't think bullets even support colliding with each other, and aiming of turrets is done by target selection which bullets cannot be). It seems that some kind of nice animation with an automatic point-defense-system would be the easiest/nicest system to implement, I think. A simple line drawing thing could do the trick nicely.

Perhaps such a point-defense laser system could even work on bullets too, or at least kinetic ones. Actually, that could create an interesting dynamic between kinetic and energy bullets with energy bullets being shorter-range in general but kinetic weapons being possible to counter with point-defense (but I figure there would be a hierarchy of "level X point defense needed to destroy bullet Y". This could be used on missiles, too.

Also, I really like the idea of making point-defense a thing mainly for medium ships (where, like you said, there wouldn't be much point in putting them on larger ships - but perhaps to a small extent they could make sense for lighter large ships like the Kestrel).

Scaling jammers that way sounds like an interesting thought too.

**Comment by bobbens** [comment]
I agree that missiles are broken as they are right now. They are basically a power boost for the NPC enemies. Some good questions are raised on what should the role of missiles be. Defining its role first would then allow us to see how to implement them. My thoughts.

* Having ammo is bad. However, if you remove ammo, they basically can wind up playing the same role as "standard weapons".
* "As equalizers for smaller ships against larger ships" is probably one of the easier to define roles. Bombers would be able to deal really high damage vs capital ships despite being very small. However, they can't lock on small ships.

What roles they can play besides bombers is something that's not clear. The only possibility I can think of is:

1. Rock-paper-scissors style with Point Blank Defenses (PBD). Have them do a lot of damage if you don't have PBD but if you have it you can nullify a certain amount. However, if they enemy doesn't have missiles, then PBD is just wasted ship potential.

Regarding some other comments.

> * Maybe missiles could be reconfigured to only be usable by medium and large ships? But that could create a bad power balance in the opposite direction.

I don't think this is a good idea.

> * Maybe missiles could have their range shortened? But that would make them a lot less useful.
>   * Maybe short-range missiles could be tweaked in such a way that they're used alongside other weapons the same way, but easier to aim?

Shorter range increases overlap with "standard weapons", so keeping range as a distinguishing feature is probably good.

> * Maybe remove the lock-on requirement?

Electronic warfare is a bit of a mess. The idea is good in my opinion, but it is both hard to balance and hard for the player to understand. However, I think it might be best to keep it as is.

> * Maybe we need some sort of "missile countermeasures" feature that costs some battery or something but destroys incoming missiles? But it's hard to say how to make a feature like that integrate into Naev well.

Point Blank Defense could be implemented as a weapon slot outfit that basically deals X damage per second to the nearest N missiles within Y range and use Z energy per second. If you give missiles HP, then it's just straightforward implementation. It would also be possible to make, for example, some anti-capship missiles targetable with normal weapotargetable so you could have fighters protect capships from bombers. This could be done with a flag and would be a bit harder to implement.

> * Maybe make missiles so slow that they're only effective on large ships? But then that would just shift the main harm missiles cause to large ships.

This role is good, but it would be nice to have other usages also.

> * Maybe give ships designated missile slots that can only be occupied by missiles, and not allow missiles to occupy any other slot (like core systems)? This of course would have to be combined with some major overhauls but maybe it could even out missile usage?

This feels like it would be limiting and also harder to balance.

That said, if removing missiles out of the equation winds up being more fun (and easier to balance), that is also a possibility. However, it does mess up stuff like the destroy the base mission, but it could be fixed with some sort of anti-base beam or something.

**Comment by onpon4** [comment]
So one thing that occurs to me about dealing damage per second is it would make it more difficult to design it so that PBD X works against missile Y but not against missile Z. Just having a rudimentary check of "does PBD X work against missile/bullet Y" would be easier to design. Especially when considering the possibility of a player merging together a jammer with a normally ineffective PBD to make it effective (because jammers make the missiles take more time to hit) - which could be a real interesting feature, but at the same time could make it difficult to prevent gaming the system while still making PBD effective *without* doing so (imagine either a capital ship being immune to missiles that it isn't supposed to be, or a medium ship without a jammer being unable to stop a certain mid-power missile). Extensive testing could weed these sorts of things out of course, but it would be a lot easier to just make it a clear-cut case of "this PBD works against these missiles/bullets".

That would also make the PBD more efficient since you wouldn't be wasting energy on missiles you can't even actually destroy.

As regards the ammo thing, I feel like ammo should be a thing, but slowly regenerating when out of combat and fully restored when you land (which would actually make it behave a lot like battery, but just at a much slower rate). I figure going through active cooldown could also restore ammo completely so you're not encouraged to sit around for several minutes at a time restocking your missiles. Basically the goal would be to limit available missiles in a given battle, but not for a given space flight.

Or perhaps there should be a different, less reliable way to refill missiles out in space. But I can't think of one that would work well with Naev's space combat style. :/

**Comment by bobbens** [comment]
> So one thing that occurs to me about dealing damage per second is it would make it more difficult to design it so that PBD X works against missile Y but not against missile Z. Just having a rudimentary check of "does PBD X work against missile/bullet Y" would be easier to design. Especially when considering the possibility of a player merging together a jammer with a normally ineffective PBD to make it effective (because jammers make the missiles take more time to hit) - which could be a real interesting feature, but at the same time could make it difficult to prevent gaming the system while still making PBD effective _without_ doing so (imagine either a capital ship being immune to missiles that it isn't supposed to be, or a medium ship without a jammer being unable to stop a certain mid-power missile). Extensive testing could weed these sorts of things out of course, but it would be a lot easier to just make it a clear-cut case of "this PBD works against these missiles/bullets".

What about some number value? Something like evasion and tracking. If your PBD can track 5.0 then anything with evasion <5.0 takes damage, and stuff with >5.0 goes through. This would allow more simple balancing rather than some sort of flag per weapon or something like that.


> As regards the ammo thing, I feel like ammo should be a thing, but slowly regenerating when out of combat and fully restored when you land (which would actually make it behave a lot like battery, but just at a much slower rate). I figure going through active cooldown could also restore ammo completely so you're not encouraged to sit around for several minutes at a time restocking your missiles. Basically the goal would be to limit available missiles in a given battle, but not for a given space flight.

Yes, that's what I always thought would make most sense. Missiles would allow you to increase your dps over standard weapons until you run out of ammo, when dps would drop significantly due to regeneration rate. This has the side effect to encourage their usage in small skirmishes rather than drawn-out fights though.

**Comment by onpon4** [comment]
Yeah, a number value was what I was thinking. Though I was thinking more of calling it something like "resilience" and "power" so that e.g. a power 5 PBD laser can destroy a resilience 5, 4, 3, 2, 1 missile, but not a resilience 6 missile. (Making it float would work just fine too, though I imagine unnecessary.) Lasers all travel at the speed of light so I'm thinking of it in terms of not so much evasion as what lasers are actually effective. So internally it would decide if the laser should win and if so do the laser strike and destroy the missile, otherwise not do the laser strike (save battery).

> This has the side effect to encourage their usage in small skirmishes rather than drawn-out fights though.

I actually rather like that effect and I think it should be expanded so that missiles don't reload _at all_ if you're in combat. But maybe they could refill to capacity once you're sufficiently out of combat? We could even go extreme on this just to make absolutely sure of this, like only refilling missiles when you jump or land for instance. But otherwise we could make it check things like battery and shield and damage and stuff to determine if the pilot is in a combat situation and restore missile ammo if not.

**Comment by bobbens** [comment]
> Yeah, a number value was what I was thinking. Though I was thinking more of calling it something like "resilience" and "power" so that e.g. a power 5 PBD laser can destroy a resilience 5, 4, 3, 2, 1 missile, but not a resilience 6 missile. (Making it float would work just fine too, though I imagine unnecessary.) Lasers all travel at the speed of light so I'm thinking of it in terms of not so much evasion as what lasers are actually effective. So internally it would decide if the laser should win and if so do the laser strike and destroy the missile, otherwise not do the laser strike (save battery).

Sounds good.

> I actually rather like that effect and I think it should be expanded so that missiles don't reload _at all_ if you're in combat. But maybe they could refill to capacity once you're sufficiently out of combat? We could even go extreme on this just to make absolutely sure of this, like only refilling missiles when you jump or land for instance. But otherwise we could make it check things like battery and shield and damage and stuff to determine if the pilot is in a combat situation and restore missile ammo if not.

I would just treat it the same way heat is treated. As in you have to use active cooldown to force it to "reload" otherwise it has a fairly slow recharge rate. This avoids adding more complications (combat detection, etc.), and is more intuitive to the player. Alternatively, if done with combat detection, it might actually be good to remove active cool down and make "heat" work the same way, i.e., cooling down really fast when not in combat.

**Comment by onpon4** [comment]
Yeah, that does seem like it would work and keep things pretty simple. OK! I think I have a pretty good idea what I want to do (assuming no one beats me to it of course).

**Comment by onpon4** [comment]
So, I've been testing out the auto-refilling missiles (#862), and I'd like to remark that even this alone has substantially improved balance issues. It's in particular been of massive help for a Rhino's defenses; it's doing very will fending off pirates with a couple of Plasma Turret MK3's and a pair of turreted Fury launchers.

I think it might even be possible to balance this right without the PBD system (which I'm still in favor of, though not planning on implementing imminently). In particular, I'm thinking:

* Straight-shot, slow, softly guided missiles and dumbfire rockets for bombers, corvettes, and light destroyers to use against mostly larger ships from a distance.
* Straight-shot, agile, guided missiles for corvettes and light destroyers to use against distant smaller ships.
* Turreted, agile, guided missiles for heavy destroyers, large civilian ships, and light cruisers to use against distant smaller ships.
* Fighter bays for armored transports, heavy cruisers, and especially carriers to use against distant smaller ships and for general defense.

Thoughts on that?

**Comment by bobbens** [comment]
> * Straight-shot, slow, softly guided missiles and dumbfire rockets for bombers, corvettes, and light destroyers to use against mostly larger ships from a distance.
> * Straight-shot, agile, guided missiles for corvettes and light destroyers to use against distant smaller ships.
> * Turreted, agile, guided missiles for heavy destroyers, large civilian ships, and light cruisers to use against distant smaller ships.
> * Fighter bays for armored transports, heavy cruisers, and especially carriers to use against distant smaller ships and for general defense.

I think it sounds good as a set of general rules, as long as there are still surprises for the player.

**Comment by onpon4** [comment]
I think I can say with confidence at this point that this is solved (by an old pull request, not sure which). I've been successfully playing with all types of ships and have done well with all of them.
